HIVE-29098 : Fix dataloss post migrating a partitioned table with default partition to iceberg

vikramahuja1001 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

deniskuzZ commented on iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:
```diff
@@ -169,7 +169,10 @@ private static List<DataFile> getDataFiles(RemoteIterator<LocatedFileStatus> fil
             StructLike structLike = dataFile.partition();
             for (int pos : stringFields) {
-              structLike.set(pos, FileUtils.unescapePathName(structLike.get(pos, String.class)));
+              String val = structLike.get(pos, String.class);
```
@vikramahuja1001 i think we can simply revert HIVE-27367,  https://github.com/apache/iceberg/pull/7744 fixed the issue. Could you please try it out.

vikramahuja1001 replied:
Thanks for the review @deniskuzZ . I will check the same and revert back.

vikramahuja1001 replied:
@deniskuzZ , simple reverting the code part of HIVE-27367 works in this case. I have reverted the code but have kept the test case.
Also added new tests as well as a part of this PR.

vikramahuja1001 commented:
Simply revert [HIVE-27367](https://issues.apache.org/jira/browse/HIVE-27367), https://github.com/apache/iceberg/pull/7744 fixes this issue. Reverted the code part of HIVE-27367 and have kept the test case to prevent any future regressions wrt HIVE-27367. Also added more tests as a part of this PR so that no regression happens in the future.
cc @deniskuzZ
